[base][docs] Base UI Quickstart Guide (#36717)
mui/material-ui: docs/data/base/getting-started/quickstart/quickstart.md

# Quickstart

<p class="description">Get started with Base UI, a library of headless ("unstyled") React UI components and low-level hooks.</p>

## Installation

`@mui/base` is completely standalone – run one of the following commands to add Base UI to your React project:

### With npm

```sh
npm install @mui/base
```

### With yarn

```sh
yarn add @mui/base
```

### With pnpm

```sh
pnpm add @mui/base
```

### Peer dependencies

<!-- #react-peer-version -->

[`react`](https://www.npmjs.com/package/react) >= 17.0.0 and [`react-dom`](https://www.npmjs.com/package/react-dom) >= 17.0.0 are peer dependencies.

## Implementing a Button

This is a quick tutorial that goes through the basics of using and styling Base UI components by replicating a button from GitHub's UI, using their [Primer design system](https://primer.style) as a reference.

> Base UI components help you abstract away low-level UI implementation details such as accessibility patterns, cross-browser compatibility, event handling, and more, while remaining unopinionated about styling, giving you complete control over your app's CSS.

Reviewer: I'd be down for placing this paragraph somewhere else, folks might be convinced of using Base UI already if they're diving into the Quickstart doc. But if we were to keep it, also down for your title suggestions there 🤙

> Base UI components help you abstract away low-level UI implementation details such as accessibility patterns, cross-browser compatibility, event handling, and more, while remaining unopinionated about styling, giving you complete control over your app's CSS.

Author: Let me take it out for now, and we can add it separately to the Overview page as Sam previously suggested !

{{"demo": "Tutorial.js", "defaultCodeOpen": false, "hideToolbar": true}}

### Components and hooks

> Base UI provides a Button component and a `useButton` hook.

Reviewer 2: ```suggestion
Base UI provides a Button component and a useButton hook.
```

or "a `Button` component and a `useButton` hook"

Reviewer 3: @michaldudak according to the [Style Guide](https://www.notion.so/mui-org/Writing-style-guide-2a957a4168a54d47b14bae026d06a24b), the way it's written currently is correct. (Of course it can always be changed as we see fit.)

If you wanted to put both in backticks, the preferred style would be `<Button />` and `useButton`. If you think that neither of them should have backticks, then there are many other instances of `useSomeHook` throughout the Base and Joy docs that we'll want to update to remove backticks.

Author: Let's go with `<Button />` here as it's the first occurrence of the button component, subsequent ones will have no backtick

# Quickstart

<p class="description">Get started with Base UI, a library of headless ("unstyled") React UI components and low-level hooks.</p>

## Installation

`@mui/base` is completely standalone – run one of the following commands to add Base UI to your React project:

### With npm

```sh
npm install @mui/base
```

### With yarn

```sh
yarn add @mui/base
```

### With pnpm

```sh
pnpm add @mui/base
```

### Peer dependencies

<!-- #react-peer-version -->

[`react`](https://www.npmjs.com/package/react) >= 17.0.0 and [`react-dom`](https://www.npmjs.com/package/react-dom) >= 17.0.0 are peer dependencies.

> ## Implementing a Button

Reviewer 4: I wonder if it is worth adding a paragraph that explains why you should use Base UI button instead of a plain HTML button.

I guess a lot of people might have this question in their minds. cc @samuelsycamore

Reviewer 3: Great idea @siriwatknp !

> This is a quick tutorial that goes through the basics of using and styling Base components by replicating a button from Github's UI, using their [style guide](https://primer.style) as a reference.

Reviewer: ```suggestion
This is a quick tutorial that goes through the basics of using and styling Base components by replicating a button from GitHub's UI, using their [Primer design system](https://primer.style) as a reference.
```

Sorry ⎯ such a pedantic change but it's important to me 😅 😆

> This is a quick tutorial that goes through the basics of using and styling Base components by replicating a button from Github's UI, using their [style guide](https://primer.style) as a reference.

Reviewer 2: ```suggestion
This is a quick tutorial that goes through the basics of using and styling Base UI components by replicating a button from Github's UI, using their [style guide](https://primer.style) as a reference.
```

{{"demo": "Tutorial.js", "defaultCodeOpen": false, "hideToolbar": true}}

### Components & Hooks

Base UI provides a ButtonUnstyled component and a useButton hook, both of which can be used to build a button, each with its own [benefits and trade-offs](/base/getting-started/usage/#components-vs-hooks).

#### ButtonUnstyled Component

```tsx
import * as React from 'react';
import ButtonUnstyled from '@mui/base/ButtonUnstyled';

export default function App() {
  return (
    <div>
      <ButtonUnstyled>ButtonUnstyled</ButtonUnstyled>
    </div>
  );
}
```

#### useButton Hook

```tsx
import * as React from 'react';
import useButton from '@mui/base/useButton';

export default function App() {
  const buttonRef = React.useRef();

  const { getRootProps } = useButton({

>     ref: buttonRef,

Reviewer 2: There's no need to pass a ref if you don't do anything with it. `useButton({})` should be enough in this case (but I just realized it looks weird and we should allow `useButton()`)

+ other places that show to use a hook

Author: 👌 Got it, I've opened some related issues:
- https://github.com/mui/material-ui/issues/36880
- https://github.com/mui/material-ui/issues/36879

import { ThemeProvider } from '@emotion/react';
import ButtonUnstyled from '@mui/base/ButtonUnstyled';

const theme = {
  colors: {
    primary: 'green',
  },
};

const GithubButton = styled(ButtonUnstyled)(
  ({ theme }) => `
    background-color: ${theme.colors.primary /* => 'green' */};
  `,
);

render(
  <ThemeProvider theme={theme}>
    <GithubButton>Create Repository</GithubButton>
  </ThemeProvider>,
);
```

Most of the demos and examples in the Base UI documentation are styled with MUI system in this way. You can inspect the `theme` object used on this site in your browser console, or explore the default structure [here](/material-ui/customization/default-theme/).

#### Styling ButtonUnstyled with MUI System

```tsx
import * as React from 'react';
import ButtonUnstyled from '@mui/base/ButtonUnstyled';
import { styled } from '@mui/system';

const GithubButton = styled(ButtonUnstyled)(
  ({ theme }) => `
    background-color: ${theme.palette.mode === 'dark' ? '#238636' : '#1f883d'};
    ${/* ... the rest of the styles */}
  `,
);

export default function App() {
  return (
    <div>
      <GithubButton>Create Repository</GithubButton>
    </div>
  );
}
```

> #### Styling useButton with MUI System

Reviewer 3: ```suggestion
#### useButton hook with MUI System
```

```tsx
import * as React from 'react';
import useButton from '@mui/base/useButton';
import { styled } from '@mui/system';

const GithubButton = styled('button')(
  ({ theme }) => `
    background-color: ${theme.palette.mode === 'dark' ? '#238636' : '#1f883d'};
    ${/* ... the rest of the styles */}
  `,
);

export default function App() {
  const { getRootProps } = useButton(/* props*/);

  return (
    <div>
      <GithubButton type="button" {...getRootProps()}>
        Create Repository
      </GithubButton>
    </div>
  );
}
```

> #### Using the sx prop

Reviewer 3: I think this is worth upgrading to an h3 so it sits at the same level as the other styling options.

```suggestion
### Styling with the sx prop
```

MUI System supports the [`sx` prop](/system/getting-started/the-sx-prop/), which provides a quick way to apply ad-hoc styles using theme-aware values to any component created with `styled`.

```tsx
const GithubButton = styled(ButtonUnstyled)(
  ({ theme }) => `
    background-color: ${theme.palette.mode === 'dark' ? '#238636' : '#1f883d'};
    margin: 0;
  `,
);

export default function App() {
  return (
    <div>
      <GithubButton sx={{ m: 2 /* => margin: 16px */ }}>
        Create Repository
      </GithubButton>
    </div>
  );
}
```

> Here's a complete demo of styling ButtonUnstyled and useButton with MUI System:

Reviewer 3: ```suggestion
The demo below shows how to build the Primer button using MUI System along with the `sx` prop:
```

{{"demo": "BaseButtonMuiSystem.js", "defaultCodeOpen": false}}

> Read more about working with MUI System and its utilities [here](/system/getting-started/usage/).

Reviewer 3: ```suggestion
Read the [MUI System Usage](/system/getting-started/usage/) doc for further details.
```

# Quickstart

<p class="description">Get started with Base UI, a library of headless ("unstyled") React UI components and low-level hooks.</p>

## Installation

`@mui/base` is completely standalone – run one of the following commands to add Base UI to your React project:

### With npm

```sh
npm install @mui/base
```

### With yarn

```sh
yarn add @mui/base
```

### With pnpm

```sh
pnpm add @mui/base
```

### Peer dependencies

<!-- #react-peer-version -->

[`react`](https://www.npmjs.com/package/react) >= 17.0.0 and [`react-dom`](https://www.npmjs.com/package/react-dom) >= 17.0.0 are peer dependencies.

## Implementing a Button

This is a quick tutorial that goes through the basics of using and styling Base UI components by replicating a button from GitHub's UI, using their [Primer design system](https://primer.style) as a reference.

> Base UI components help you abstract away low-level UI implementation details such as accessibility patterns, cross-browser compatibility, event handling etc while remaining unopinionated about styling, giving you complete control over your app's CSS.

Author: > I wonder if it is worth adding a paragraph that explains why you should use Base UI button instead of a plain HTML button.

@samuelsycamore I've added a sentence here, not sure if it's the smoothest transition 😅

Reviewer: Should we add a heading to separate these two? Quite directly "Why use Base UI's button instead of native HTML?" or something like that. 😬

Reviewer 3: @danilo-leal I think that's a great idea, but I wonder if this is the best place for it? I would assume that if you've made it to the Quickstart, you've already concluded that you want to use Base's Button because it's better than an HTML button. 😁 Maybe this calls for an update to the Overview page?

Reviewer 3: Avoid Latinisms whenever possible 😇

```suggestion
Base UI components help you abstract away low-level UI implementation details such as accessibility patterns, cross-browser compatibility, event handling, and more, while remaining unopinionated about styling, giving you complete control over your app's CSS.
```

Reviewer: Oh, yeah ⎯ that checks out! Might make sense to tackle it together with the "pro vs. con" thread that was going on. Base UI should add some advantage against native HTML 😬

{{"demo": "Tutorial.js", "defaultCodeOpen": false, "hideToolbar": true}}

### Components and hooks

Base UI provides an Unstyled Button component and a `useButton` hook.

> Both can be used to build a button, and each has its own benefits and trade-offs — see [Components vs. hooks](/base/getting-started/usage/#components-vs-hooks) for details.

Reviewer 3: No spaces on either side of an em dash (—).

```suggestion
Both can be used to build a button, and each has its own benefits and trade-offs—see [Components vs. hooks](/base/getting-started/usage/#components-vs-hooks) for details.
```

The code snippets below demonstrate the basic implementation of each:

#### Unstyled Button component

```tsx
import * as React from 'react';
import ButtonUnstyled from '@mui/base/ButtonUnstyled';

export default function App() {
  return (
    <ButtonUnstyled>ButtonUnstyled</ButtonUnstyled>
  );
}
```

#### useButton hook

```tsx
import * as React from 'react';
import useButton from '@mui/base/useButton';

export default function App() {
  const { getRootProps } = useButton();
  return (
    <button type="button" {...getRootProps()}>
      useButton
    </button>
  );
}
```

> Base UI comes with no styles or styling solution - here's what the Unstyled Button looks like out of the box:

Reviewer 3: Em dash.

```suggestion
Base UI comes with no styles or styling solution—here's what the Unstyled Button looks like out of the box:
```
